File: frontend/src/services/security.ts
import api from './api';

export const securityService = {
    async listRules(serverId?: string) {
        const params = serverId ? { server_id: serverId } : {};
        const { data } = await api.get('/security/rules', { params });
        return data;
    },
    async createRule(payload: {
        server_id: string; port: string; protocol?: string;
        action?: string; source_ip?: string; description?: string;
    }) {
        const { data } = await api.post('/security/rules', payload);
        return data;
    },
    async deleteRule(id: string) {
        await api.delete(`/security/rules/${id}`);
    },
    async listEvents(page = 1, perPage = 50) {
        const { data } = await api.get('/security/events', { params: { page, per_page: perPage } });
        return data;
    },
};

File: backend/internal/handlers/security.go
package handlers

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/novapanel/novapanel/internal/models"
	"github.com/novapanel/novapanel/internal/services"
)

type SecurityHandler struct {
	service *services.SecurityService
}

func NewSecurityHandler(service *services.SecurityService) *SecurityHandler {
	return &SecurityHandler{service: service}
}

func (h *SecurityHandler) CreateRule(c *gin.Context) {
	var req models.CreateFirewallRuleRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, models.ErrorResponse{Error: err.Error()})
		return
	}
	rule, err := h.service.CreateRule(c.Request.Context(), req)
	if err != nil {
		c.JSON(http.StatusBadRequest, models.ErrorResponse{Error: err.Error()})
		return
	}
	c.JSON(http.StatusCreated, rule)
}

func (h *SecurityHandler) ListRules(c *gin.Context) {
	serverID := c.Query("server_id")
	rules, err := h.service.ListRules(c.Request.Context(), serverID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{Error: err.Error()})
		return
	}
	c.JSON(http.StatusOK, rules)
}

func (h *SecurityHandler) DeleteRule(c *gin.Context) {
	if err := h.service.DeleteRule(c.Request.Context(), c.Param("id")); err != nil {
		c.JSON(http.StatusNotFound, models.ErrorResponse{Error: err.Error()})
		return
	}
	c.JSON(http.StatusOK, models.SuccessResponse{Message: "Rule deleted"})
}

func (h *SecurityHandler) ListEvents(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	perPage, _ := strconv.Atoi(c.DefaultQuery("per_page", "50"))
	resp, err := h.service.ListEvents(c.Request.Context(), page, perPage)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{Error: err.Error()})
		return
	}
	c.JSON(http.StatusOK, resp)
}
